Add Metadata classes

CrazyIvan359 opened the pull request:
As I mentioned, metadata classes!

@openhab-5iver I don't have write access on #61 so I couldn't push to that branch. I figured since mine differs so much from that one I would start my own branch and PR for it.

@mjcumming Credit where credit is due Michael, I started writing this module with your code in #61 as a reference point and it was a great help as I had not done anything with metadata before this. I found your class to be rather raw in that you always manually read and write to the registry. I moved to a more hands off approach where the class immediately loads when created and writes when any changes are made. My `metadata_namespace` class loosely resembles your class, and I added another simple class to allow managing multiple namespaces of a single item from one object.

&nbsp;
> and a provider should be used (which I think I wrote already).

I forgot you had mentioned this until I was looking at the other PR to compose this message. I don't see anything metadata in the codebase. 
Presently I am not using a Provider, but I found the code in OH Core. I will look into this more, though at present I'm not entirely sure it's necessary for the purposes of this class. As I understand it, the Provider allows OH to know what attributes a namespace should have, but that seems beyond the scope of this module as it stands.
I could create a Provider that would allow adding namespace definitions from JSR223 code, this would be useful for what I am getting into as it would allow changes to the namespaces from PaperUI.

CrazyIvan359 commented:
> Presently I am not using a Provider, but I found the code in OH Core. I will look into this more, though at present I'm not entirely sure it's necessary for the purposes of this class. As I understand it, the Provider allows OH to know what attributes a namespace should have, but that seems beyond the scope of this module as it stands.
I could create a Provider that would allow adding namespace definitions from JSR223 code, this would be useful for what I am getting into as it would allow changes to the namespaces from PaperUI.

So I dug around in openhab-core and understand Providers/Registries a little more now. They do not provide templates, but I'm still not sure they are the right approach for this class.
If I use a custom provider we would only have access to metadata that had been created in Jython code. The `MetadataRegistry` seems to implement a built in `ManagedMetadataProvider` that looks after any add/update/remove requests and saves to the jsondb. While this approach makes sense for most use cases when thinking of items and thing creation in scripts, I don't think its the best approach for metadata.

mjcumming commented:
@CrazyIvan359 changes to metadata look great.

openhab-5iver commented:
Once lucid is merged, lets get this one merged and then go after the documentation.

mjcumming commented:
Let me know what is needed

CrazyIvan359 commented:
@openhab-5iver I think this is ok to merge, let me know if you need me to make any changes though

openhab-5iver commented on Core/automation/lib/python/core/metadata.py:
```diff
@@ -0,0 +1,193 @@
+            item=item_name, namespace=name))
+        self._item_name = item_name
+        self._name = name
+        self._keyUID = MetadataKey(self._name, self._item_name)
```
_key_uid

CrazyIvan359 replied:
Oops right beside some lower_case ones too lol. I think I did this while testing and reflexively used camel case since it's my go to

openhab-5iver commented on Core/automation/lib/python/core/metadata.py:
```diff
@@ -0,0 +1,193 @@
+    return False if MetadataRegistry.get(MetadataKey(namespace, item_name)) is None else True
+
+def _resolveType(value):
+    '''Attempts to resolve the type of the metadata value.
```
My preference is to start and end block comments on their own lines too. I think it is easier to read and identify as a comment. WDYT? Although, if we [stuck to the standard](https://www.python.org/dev/peps/pep-0008/#block-comments), we'd be using #s.

CrazyIvan359 replied:
Like this?
```python
def _resolve_type(value):
    '''
    Attempts to resolve the type of the metadata value.
    '''
````

I have no preference really. I tend to have them on the same line to save on line count. I use VSC so with the highlighting my eyes just ignore quotes when they see green.

openhab-5iver replied:
> Like this?

Yes, that is what I was meaning. Visually, it's more 'blocky', which IMO is easier to read. I use VSC too, but it was tough to read it in GH!

CrazyIvan359 replied:
That's fair, I hadn't really noticed on GH

openhab-5iver commented on Core/automation/lib/python/core/metadata.py:
```diff
@@ -0,0 +1,193 @@
+        else:
+            if remove: self._namespaces[name].remove
+            self._namespaces.pop(name, None)
+
```
extra empty line

CrazyIvan359 replied:
Like this for readability?
```python
        else:
            if remove: self._namespaces[name].remove
            
            self._namespaces.pop(name, None)
```

openhab-5iver replied:
No, no. :slightly_smiling_face: There are two empty lines there, so I was suggesting to remove one.

CrazyIvan359 replied:
Oh I see, I wondered about that while I was working on it last night. I usually do that between classes/functions and single line spaces between methods or subfuncs so it's easier to see the separation. I was inconsistent with that in the file though. So a preference for single blank lines throughout then?

openhab-5iver replied:
Yes, but I'm not sure if PEP8 says anything about it.

CrazyIvan359 replied:
PEP8 says what I do, but should we stick to one blank for consistency?

> Surround top-level function and class definitions with two blank lines.
>
> Method definitions inside a class are surrounded by a single blank line.
>
> Extra blank lines may be used (sparingly) to separate groups of related functions. Blank lines may be omitted between a bunch of related one-liners (e.g. a set of dummy implementations).
>
> Use blank lines in functions, sparingly, to indicate logical sections.

openhab-5iver commented on Core/automation/lib/python/core/metadata.py:
```diff
@@ -0,0 +1,193 @@
+        except ValueError: pass
+        # attempt to cast to float
+        try: return float(str(value))
+        except ValueError: pass
```
Sorry if this is nit picky, but I think these should go on separate lines for readability. There are some other places too.

CrazyIvan359 replied:
Consistency is important. This is my style, not the repo's. Thinking about it, I don't know where I picked up the habit of trying to keep the line count down

openhab-5iver replied:
I tend to do the same. In the past, I would try to condense everything as much as I could, but I'm becoming more used to the concept of readability.

openhab-5iver commented on Docs/Jython-Modules.md:
```diff
@@ -197,6 +197,43 @@ core.link.remove_link("Kodi_Control")
+#### Module: [`core.metadata`](..Core/automation/lib/python/core/metadata.py)
+<ul>
+
+This module allows simpler interaction with item metadata. There are two instance classes that can be used manage a single namespace, or multiple namespaces of a single item.
```
In an md, put sentences on separate lines with a space after the period.

CrazyIvan359 replied:
I'll make the changes, but purely out of curiosity, why?

openhab-5iver replied:
If there's one big blob of text on a line, it's harder to track changes. openhab-docs requires per sentence tracking too. I think there might even be a SAT check for this.

CrazyIvan359 commented:
> This looks well written. Have you fully tested it though? I had to do a PR recently for core.date, due to a number of errors we came across in the lucid migration.

I have not fully tested it yet, no. I also forgot that I never did tests of all of `core.date` either. I don't have a working dev setup for openHAB + Jython right now though, thanks to the ESH merge, so testing is a bit difficult for me. Do you know of a way to `apt` install _any_ build of OH? I can only see about a dozen recent builds and don't think I found an archive of build packages anywhere. Even better would be your branch that fixes all of the import names post ESH merge. I haven't had much time to work on a computer in the last couple weeks, though I might have enough time to test tonight.

When I wrote this, I did use it to read and correctly type all types it recognises. It reads value and the config array correctly. I only used the namespace class, not the item class in those tests.

> Some simple examples at the top of the Jython-Modules.md entry would be helpful... like adding, reading, removing metadata from an Item.

Good idea. I'll clean it up.
  
  
I will finish up the page I'm writing in the docs and get to work on this

CrazyIvan359 commented on Core/automation/lib/python/core/metadata.py:
```diff
@@ -0,0 +1,193 @@
+from core.log import logging, LOG_PREFIX
+log = logging.getLogger(LOG_PREFIX + ".metadata")
+from core import osgi
+MetadataRegistry = osgi.get_service("org.eclipse.smarthome.core.items.MetadataRegistry")
```
Should we not keep this name as is since it isn't ours? I used this case because its a name from OH

openhab-5iver replied:
Actually, I just found that the precedent has already been set in core.actions to use snake_case. So, something like metadata_registry.

CrazyIvan359 commented:
I've made some of the changes, waiting on your input on others. Also cleaned up a few things, added some sanity checks on namespace names, checking for item existing, and added `None` to recognised types.

I did not have time to setup a test env to do further tests yet and may not have time to work on this tomorrow, so I've pushed the changes here in case you wanted to work on it.

openhab-5iver commented on Core/automation/lib/python/core/metadata.py:
```diff
@@ -0,0 +1,297 @@
+        '''
+        Sets the namespace "value" and saves to the registry.
+        '''
+        return self.set_value(value)
```
Testing the first example (metadata_namespace), I get an error when using the module. Removing this function fixed it, but I did not test all methods yet. There is another function named `value` at L233, but I think this issue was with `value.setter`. The error was...

```
2019-05-12 10:59:48.624 [ERROR] [org.openhab.core.automation.module.script.internal.ScriptEngineManagerImpl] - Error during evaluation of script 'file:/opt/openhab2/conf/automation/jsr223/python/personal/test/testScript.py': NameError: name 'value' is not defined in <script> at line number 53
```

CrazyIvan359 replied:
I think this is one I added after I had done my tests. I think I know what it's complaining about though. When defining custom properties as methods like that you have to start with the getter then have the setter. If memory serves (I'm on my phone) it's like this:

```python
@property
def value(self):
    return self._value

@value.setter
def value(self, value):
    self._value = value
```

I may have forgotten to add that getter. I'll look at it tonight

CrazyIvan359 replied:
Oops, you said there is another at line 233, that's the issue, the getter needs to be first

CrazyIvan359 commented on Core/automation/lib/python/core/metadata.py:
```diff
@@ -188,23 +189,33 @@ def save(self):
-            strValue = "None"
-        else:
-            strValue = str(self._value)
+        strValue = str(self._value)
```
Turns out `str(None)` gets you `'None'` so I removed this extra check

CrazyIvan359 commented:
@openhab-5iver was this just waiting on testing?

openhab-5iver commented:
Documentation, mainly examples, and testing. It was difficult for me to use it. I also question why we'd want the caching, which complicates things. This one is next, after the doc rewrite.

mjcumming commented:
> Documentation, mainly examples, and testing. It was difficult for me to use it. I also question why we'd want the caching, which complicates things. This one is next, after the doc rewrite.

I have kept with the implementation I did a pull request for - its not as versatile but its simpler and fits better with my needs.

CrazyIvan359 commented:
Well the caching is for simplicity I suppose, reading from the registry is a bit involved. Also consider the life-cycle of the class, it will only be around for the time it takes a rule to run.

mjcumming commented:
> Well the caching is for simplicity I suppose, reading from the registry is a bit involved. Also consider the life-cycle of the class, it will only be around for the time it takes a rule to run.

I depends on your use case - I have scripts that constantly refer to metadata  and change it as required. Caching would introduce some potential issues that would be hard to track down.

CrazyIvan359 commented:
This module has undergone major changes and a new PR has been made (#167)
